scripts/base/protocols/websocket/main.zeek

```diff
@@ -0,0 +1,219 @@
+			{
+			if ( ! ws?$client_extensions )
+				ws$client_extensions = vector();
+
+			ws$client_extensions += split_string(value, / *, */);
+			}
+		else if ( name == "SEC-WEBSOCKET-KEY" )
+			{
+			if ( ws?$client_key )
+				Reporter::conn_weird("websocket_multiple_key_headers", c, "", "WebSocket");
+
+			ws$client_key = value;
+			}
+		}
+	else
+		{
+		if ( name == "SEC-WEBSOCKET-PROTOCOL" )
+			{
+			if ( ws?$subprotocol )
+				Reporter::conn_weird("websocket_multiple_protocol_headers", c, "", "WebSocket");
+
+			ws$subprotocol = value;
+			}
+		else if ( name == "SEC-WEBSOCKET-EXTENSIONS" )
+			{
+			if ( ! ws?$server_extensions )
+				ws$server_extensions = vector();
+
+			ws$server_extensions += split_string(value, / *, */);
+			}
+		else if ( name == "SEC-WEBSOCKET-ACCEPT" )
+			{
+			if ( ws?$server_accept )
+				Reporter::conn_weird("websocket_multiple_accept_headers", c, "", "WebSocket");
+
+			ws$server_accept = value;
+			}
+		}
+	}
+
+event http_request(c: connection, method: string, original_URI: string,
+                   unescaped_URI: string, version: string)
+	{
+	# If we see a http_request and have websocket state, wipe it as
+	# we should've seen a websocket_established even on success and
+	# likely no more http events.
+	if ( ! c?$websocket )
+		delete c$websocket;
+	}
+
+event websocket_established(c: connection, aid: count) &priority=5
+	{
+	if ( ! c?$websocket )
+		{
+		# This means we never saw a Sec-WebSocket-* header, weird.
+		Reporter::conn_weird("websocket_established_unexpected", c, "", "WebSocket");
+		set_websocket(c);
+		}
+
+	local ws = c$websocket;
+
+	if ( ! ws?$client_key )
+		Reporter::conn_weird("websocket_missing_key_header", c, "", "WebSocket");
+
+	if ( ! ws?$server_accept )
+		Reporter::conn_weird("websocket_missing_accept_header", c, "", "WebSocket");
+
+	# Verify the Sec-WebSocket-Accept header's value given the Sec-WebSocket-Key header's value.
+	if ( ws?$client_key && ws?$server_accept )
+		{
+		local expected_accept = expected_accept_for(ws$client_key);
+		if ( ws$server_accept != expected_accept )
+			Reporter::conn_weird("websocket_wrong_accept_header", c,
+			                     fmt("expected=%s, found=%s", expected_accept, ws$server_accept),
+			                     "WebSocket");
+		}
+
+	# Replicate some information from the HTTP.log
+	if ( c?$http )
+		{
+		if ( c$http?$host )
+			ws$host = c$http$host;
+
+		if ( c$http?$uri )
+			ws$uri = c$http$uri;
+
+		if ( c$http?$user_agent )
+			ws$user_agent = c$http$user_agent;
+		}
+	}
+
+event websocket_established(c: connection, aid: count) &priority=-5
+	{
+	local ws = c$websocket;
+
+	local config = AnalyzerConfig();
+	if ( ws?$subprotocol )
+		config$subprotocol = ws$subprotocol;
+
+	if ( ws?$server_extensions )
+		config$server_extensions = ws$server_extensions;
+
+	# Give other scripts a chance to modify the analyzer configuration.
+	hook WebSocket::configure_analyzer(c, aid, config);
+
+	WebSocket::__configure_analyzer(c, aid, config);
```

Comment:
This may happen if an `InstantiateAnalyzer` failed internally, for example, but than that already warned. And if a requested analyzer was disabled, we shouldn't spuriously log, so not sure what to do here otherwise.